src/module.c

```diff
@@ -3991,6 +3998,14 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
     RedisModuleCallReply *reply = NULL;
     int replicate = 0; /* Replicate this command? */
 
+    /* We need to use a global replication_allowed flag in order to prevent
+     * replication of nested RM_Calls. Example:
+     * 1. module1.foo does RM_Call of module2.bar without replication (i.e. no '!')
+     * 2. module2.bar internally calls RM_Call of INCR with '!'
+     * 3. at the end of module1.foo we call RM_ReplicateVerbatim
+     * We want the replica/AOF to see only module1.foo and not the INCR from module2.bar */
+    static int replication_allowed = 1;
```

Comment:
what if the nested module uses RM_Replicate rather than RM_Call with `!`?
i think we need to convert this global from the static to a real global, and prevent all propagations (at least module triggered), from happening inside RM_Call

Reply:
the RM_Replicate works as expected since it adds to the also_propagate array which is ignored because the first RM_Call didn;t have the '!'
but yeah, i guess it's safer to have a real server global and just use it in propagate()
also the diff will be smaller

Reply:
oh but is see what you mean: RM_Call without `!` then RM_Call with `!` then RM_Replicate
should i add a test?

Reply:
yes, let's add a test. as many as we can..
i think that any code running inside RM_Call should never replicate, the module that did the RM_Call is responsible for that.